### PR TITLE
feat(preview): make UI changes and allow individual dimensions to be editable

### PR DESCRIPTION
**Pull request by gargi-bansal**
### Purpose

User can now edit slider in preview page and click on dimension title to go to fully edit the dimension

[HRT-62](https://softeng761team5.atlassian.net/browse/HRT-62)

### Approach
Opens up a new component that looks identical to Display Cards to allow for full editing

### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

#### Does this PR involve a UI change?

- [x] Yes
- [ ] No

#### Screenshot(s) of UI before this PR
N/A

#### Screenshot(s) of UI after this PR
Screen Shot 2020-10-01 at 7 38 57 PM
Screen Shot 2020-10-01 at 7 38 46 PM

### CODE REVIEW

**Review by mitchfaulcon — COMMENTED** [review]
Looks good, but just checking:
 - Should the right footer button still say save when you've clicked back to a dimension from the preview page?
- Should clicking back on the preview page take you all the way back to the start or just back to the last dimension? At the moment you go all the way back to the start and have to click through them all again

**Comment by gargi-bansal** [comment]
> Looks good, but just checking:
> 
> * Should the right footer button still say save when you've clicked back to a dimension from the preview page?
> * Should clicking back on the preview page take you all the way back to the start or just back to the last dimension? At the moment you go all the way back to the start and have to click through them all again

I agree with the first one, I couldn't figure out what to name the buttons on the individual preview editing.
I think I have an idea to fix the second bullet point so I'll do it now.

File: src/components/PreviewDimension.tsx [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  previewSliderPosChange: (value: number, dimensionKey: string) => void;
+  saveDimensionFunction: (updatedDimension: DimensionType) => void;
+  openSingleDimension: (key: string) => void;
+  saveDimensionClicked: boolean;
+}
+
+export enum CardSide {
+  Left,
+  Right,
+}
```
Comment (mitchfaulcon) [inline src/components/PreviewDimension.tsx]: I think this should be imported from DisplayCards instead defined again

Comment (gargi-bansal) [reply src/components/PreviewDimension.tsx]: okay, I'll just refactor out all the common things tomorrow then!

**Review by mitchfaulcon — APPROVED** [review]

**Review by michellemtan — COMMENTED** [review]
image

Not sure if this should be part of this PR, but currently when the user hasn't selected any cards, the Preview page is blank

But when the user does the following:
1. Goes to /preview page by url and presses back button
2. Currently completed cards: 0/10, selects a card, currently completed cards updates to: 1/10
3. Next button doesn't work but back button acts as a next button?

**Comment by gargi-bansal** [comment]
> image
> 
> Not sure if this should be part of this PR, but currently when the user hasn't selected any cards, the Preview page is blank
> 
> But when the user does the following:
> 
> 1. Goes to /preview page by url and presses back button
> 2. Currently completed cards: 0/10, selects a card, currently completed cards updates to: 1/10
> 3. Next button doesn't work but back button acts as a next button?

@michellemtan I can try make it automatically redirect to another page if no cards are visible (atm I'm only showing the cards that have been filled in). Tbh I can just show all of them if that's better and have them (greyed out/disabled) except I think the slider position always sets itself to 0 for those which is incorrect, hence why I filtered them out.

The second issue about going back is actually working properly. Going back on the preview page takes you back to the last card in DisplayCards. So you should only be able to go back from there. Imagine if you had filled out all the cards and navigated to preview and then want to go back again. I think it's best for the last card to be displayed.

**Comment by mitchfaulcon** [comment]
@gargi-bansal If they directly navigate to the /Preview URL, I don't think the chart would be associated with any course so would it be best to redirect to the dashboard?

**Comment by gargi-bansal** [comment]
@mitchfaulcon I sort of want to add this as a bug on the backlog, merge my branch in and wait for the UI/backend linking to be done first. That way I can see what we store in state and how it would be easiest to detect whether we already have a chart. Atm I don't know if we're using state or will be putting the chart ID in the URL.

**Comment by mitchfaulcon** [comment]
> @mitchfaulcon I sort of want to add this as a bug on the backlog, merge my branch in and wait for the UI/backend linking to be done first. That way I can see what we store in state and how it would be easiest to detect whether we already have a chart. Atm I don't know if we're using state or will be putting the chart ID in the URL.

Sounds good to me

**Comment by michellemtan** [comment]
> > image
> > Not sure if this should be part of this PR, but currently when the user hasn't selected any cards, the Preview page is blank
> > But when the user does the following:
> > 
> > 1. Goes to /preview page by url and presses back button
> > 2. Currently completed cards: 0/10, selects a card, currently completed cards updates to: 1/10
> > 3. Next button doesn't work but back button acts as a next button?
> 
> @michellemtan I can try make it automatically redirect to another page if no cards are visible (atm I'm only showing the cards that have been filled in). Tbh I can just show all of them if that's better and have them (greyed out/disabled) except I think the slider position always sets itself to 0 for those which is incorrect, hence why I filtered them out.
> 
> The second issue about going back is actually working properly. Going back on the preview page takes you back to the last card in DisplayCards. So you should only be able to go back from there. Imagine if you had filled out all the cards and navigated to preview and then want to go back again. I think it's best for the last card to be displayed.

Re the first issue: Either works! I think it's just better than having an empty page because I thought something was broken 👍 
Re the second issue: Ohhh okay that makes sense, I think once we get the proper dimensions and cards from Claire, it'll be more obvious to navigate backwards!

**Comment by gargi-bansal** [comment]
> > > image
> > > Not sure if this should be part of this PR, but currently when the user hasn't selected any cards, the Preview page is blank
> > > But when the user does the following:
> > > 
> > > 1. Goes to /preview page by url and presses back button
> > > 2. Currently completed cards: 0/10, selects a card, currently completed cards updates to: 1/10
> > > 3. Next button doesn't work but back button acts as a next button?
> > 
> > 
> > @michellemtan I can try make it automatically redirect to another page if no cards are visible (atm I'm only showing the cards that have been filled in). Tbh I can just show all of them if that's better and have them (greyed out/disabled) except I think the slider position always sets itself to 0 for those which is incorrect, hence why I filtered them out.
> > The second issue about going back is actually working properly. Going back on the preview page takes you back to the last card in DisplayCards. So you should only be able to go back from there. Imagine if you had filled out all the cards and navigated to preview and then want to go back again. I think it's best for the last card to be displayed.
> 
> Re the first issue: Either works! I think it's just better than having an empty page because I thought something was broken 👍
> Re the second issue: Ohhh okay that makes sense, I think once we get the proper dimensions and cards from Claire, it'll be more obvious to navigate backwards!

@michellemtan As mentioned in the previous comment, I'd rather add this as a bug and wait till the backend is linked up before fixing this. Knowing what is stored in state/URL will make it easier to figure out how to solve the issue tidily.

**Comment by michellemtan** [comment]
> > > > image
> > > > Not sure if this should be part of this PR, but currently when the user hasn't selected any cards, the Preview page is blank
> > > > But when the user does the following:
> > > > 
> > > > 1. Goes to /preview page by url and presses back button
> > > > 2. Currently completed cards: 0/10, selects a card, currently completed cards updates to: 1/10
> > > > 3. Next button doesn't work but back button acts as a next button?
> > > 
> > > 
> > > @michellemtan I can try make it automatically redirect to another page if no cards are visible (atm I'm only showing the cards that have been filled in). Tbh I can just show all of them if that's better and have them (greyed out/disabled) except I think the slider position always sets itself to 0 for those which is incorrect, hence why I filtered them out.
> > > The second issue about going back is actually working properly. Going back on the preview page takes you back to the last card in DisplayCards. So you should only be able to go back from there. Imagine if you had filled out all the cards and navigated to preview and then want to go back again. I think it's best for the last card to be displayed.
> > 
> > 
> > Re the first issue: Either works! I think it's just better than having an empty page because I thought something was broken 👍
> > Re the second issue: Ohhh okay that makes sense, I think once we get the proper dimensions and cards from Claire, it'll be more obvious to navigate backwards!
> 
> @michellemtan As mentioned in the previous comment, I'd rather add this as a bug and wait till the backend is linked up before fixing this. Knowing what is stored in state/URL will make it easier to figure out how to solve the issue tidily.

My bad, I didn't look at your reply to Mitchell's comment! Sounds good to me

**Review by michellemtan — APPROVED** [review]

**Pull request merged**